tools/serverpod_cli/test/analyzer/entities/entity_analyzer/relation_keyword_test.dart

```diff
@@ -5,6 +5,160 @@ import 'package:serverpod_cli/src/util/protocol_helper.dart';
+      protocol.yamlSourceUri.path,
+      collector,
+      definition,
+      [definition],
+    );
+
+    var classDefinition = definition as ClassDefinition;
+
+    test('then no errors are collected.', () {
+      expect(collector.errors, isEmpty);
+    });
+    test('then the table is not set on relation field.', () {
+      var parent = classDefinition.findField('parent');
+      expect(parent?.parentTable, null);
+    });
+
+    test('then the scope is set to api on the relation field.', () {
+      var parent = classDefinition.findField('parent');
+      expect(parent?.scope, SerializableEntityFieldScope.api);
+    });
+
+    test(
+        'then a scalar field with the same name appended with Id is set on the relation field.',
+        () {
+      var parent = classDefinition.findField('parent');
+      expect(parent?.scalarFieldName, 'parentId');
+    });
+
+    test('then the class has a scalar field for the id.', () {
+      var parentIdField = classDefinition.findField('parentId');
+
+      expect(
+        parentIdField,
+        isNotNull,
+        reason: 'Expected to find a field named parentId.',
+      );
+    });
+
+    test('then the scalar field has the global scope.', () {
+      var parent = classDefinition.findField('parentId');
+      expect(parent?.scope, SerializableEntityFieldScope.all);
+    });
+
+    test('then the scalar field type defaults to none nullable.', () {
+      var parent = classDefinition.findField('parentId');
+      expect(
+        parent?.type.nullable,
+        isFalse,
+        reason: 'Expected to be non-nullable.',
+      );
+    });
+
+    test(
+        'then the scalar field has the parent table set from the object reference.',
+        () {
+      var parent = classDefinition.findField('parentId');
+      expect(parent?.parentTable, 'example');
+    });
+  });
+
+  group(
+      'Given a class with a self relation on a field with the class datatype where the relation is optional',
+      () {
+    var collector = CodeGenerationCollector();
+
+    var protocol = ProtocolSource(
+      '''
+class: Example
+table: example
+fields:
+  parent: Example?, relation(optional)
+''',
+      Uri(path: 'lib/src/protocol/example.yaml'),
+      [],
+    );
+
+    var definition = SerializableEntityAnalyzer.extractEntityDefinition(
+      protocol,
+    );
+
+    SerializableEntityAnalyzer.validateYamlDefinition(
+      protocol.yaml,
+      protocol.yamlSourceUri.path,
+      collector,
+      definition,
+      [definition!],
+    );
+
+    var classDefinition = definition as ClassDefinition;
+
+    test('then no errors are collected.', () {
+      expect(collector.errors, isEmpty);
+    });
+    test('then scalar field is nullable.', () {
+      var parent = classDefinition.findField('parentId');
+      expect(parent?.type.nullable, isTrue, reason: 'Expected to be nullable.');
+    });
+  });
+
+  test(
+      'Given a class with a field without a relation, then no scalar field reference is set.',
```

Comment:
Suggestion: Add same test but for a custom class without relation.

It would be good to assert that we still support custom classes on fields without relations and not only types.
If this is already covered somewhere, feel free to ignore.